README.md
# MIBRR: Multiple Imputation with Bayesian Regularized Regression
This repository hosts development of the R package `MIBRR`.

- Licensing information is given in the [LICENSE][] file.
- Built tarballs of the `MIBRR` package are available in the [builds][] 
  directory.
- Stand-alone documentation is available in the [documentation][docs] directory.
- The source files for the most recent stable version of `MIBRR` are available 
  in the [source][src] directory.

`MIBRR` is alpha software, so please expect frequent---and dramatic---changes to 
the package's functionality and user interface. Please report any bugs that you 
encounter in the issues section of the project page. You may also leave requests 
for new features in the issues section.

Thank you for your interest in the MIBRR project! I hope you find my software
useful!

## Installation
The best way to install `MIBRR` is to use the `devtools::install_github` 
function.

1. First, make sure that you have `devtools` installed on your system
2. Next, execute the following lines:

		library(devtools)
		install_github("kylelang/MIBRR/source/MIBRR")
    
3. Finally, load `MIBRR` and enjoy:

		library(MIBRR)

If the `devtools`-based approach does not work, you can download one of the
built tar-balls from the [builds][] directory and manually install the package
from source by executing the following lines:

	install.packages(pkgs  = "/save_path/MIBRR_version.tar.gz",
	                 repos = NULL,
                     type  = "source")

Where *save_path* is replaced by the (relative or absolute) file path to the
location where you saved the tar-ball, and *version* is replaced with the correct
version number for the tar-ball that you downloaded.

## Examples

The `MIBRR` package contains four primary functions: `miben`, `mibl`, `ben`, and 
`bl`.

- The underlying model in each of these four primary functions can be estimated
  using either Markov Chain Expectation Maximization (MCEM) or fully Bayesian
  modeling.
- The `miben` and `mibl` functions do multiple imputation using the Bayesian 
  elastic net and Bayesian LASSO, respectively.
  
    - A list of imputed datasets can be generated using the `getImpData` 
	  function.
	
- Basic missing data treatments using `miben` or `mibl` might look like the 
  following:

		## Load some data:
		data(mibrrExampleData)

		## Estimate the imputation models using MCEM:
		mibenOut <- miben(data       = mibrrExampleData,
                          targetVars = c("y", paste0("x", c(1 : 3))),
                          ignoreVars = "idNum")
			  
		miblOut <- mibl(data       = mibrrExampleData,
                        targetVars = c("y", paste0("x", c(1 : 3))),
                        ignoreVars = "idNum")
						
		## Estimate the imputation models using fully Bayesian modeling:
		mibenOut <- miben(data         = mibrrExampleData,
		                  targetVars   = c("y", paste0("x", c(1 : 3))),
                          ignoreVars   = "idNum",
                          doMcem       = FALSE,
                          sampleSizes  = c(500, 500),
                          lam1PriorPar = c(1.0, 0.1),
                          lam2PriorPar = c(1.0, 0.1)
                          )
			  
		miblOut <- mibl(data         = mibrrExampleData,
		                targetVars   = c("y", paste0("x", c(1 : 3))),
                        ignoreVars   = "idNum",
                        doMcem       = FALSE,
                        sampleSizes  = c(500, 500),
                        lam1PriorPar = c(1.0, 0.1)
                        )
				
		## Extract list of 100 imputed datasets:
		mibenImps <- getImpData(mibrrFit = mibenOut, nImps = 100)
		miblImps  <- getImpData(mibrrFit = miblOut, nImps = 100)
		
- The `ben` and `bl` functions fit Bayesian elastic net and Bayesian LASSO
  models to incomplete data without returning and imputed datasets.
- Use the following to fit models using `ben` or `bl`:

		## Load some data:
		data(predictData)

		trainData <- predictData$train
		testData  <- predictData$test
		
		## Estimate a Bayesian elastic net model using MCEM:
		benOut <- ben(data = trainData,
                      y    = "agree",
                      X    = setdiff(colnames(trainData), "agree")
                      )
		   
		## Estimate a Bayesian LASSO model using MCEM:
		blOut <- bl(data = trainData,
                    y    = "agree",
                    X    = setdiff(colnames(trainData), "agree")
                    )

		## Estimate a Bayesian elastic net model using full Bayes:
		benOut <- ben(data         = trainData,
                      y            = "agree",
                      X            = setdiff(colnames(trainData), "agree"),
                      doMcem       = FALSE,
                      sampleSizes  = c(500, 500),
                      lam1PriorPar = c(1.0, 0.1),
                      lam2PriorPar = c(1.0, 0.1)
                      )
		   
		## Estimate a Bayesian LASSO model using full Bayes:
		blOut <- bl(data         = trainData,
                    y            = "agree",
                    X            = setdiff(colnames(trainData), "agree"),
                    doMcem       = FALSE,
                    sampleSizes  = c(500, 500),
                    lam1PriorPar = c(1.0, 0.1)
                    )
					
		## Extract posterior parameter samples:
		benPars <- getParams(mibrrFit = benOut, target = "agree")
		blPars  <- getParams(mibrrFit = blOut, target = "agree")
		
		## Generate out-of-sample predictions:
	    benPred <- postPredict(mibrrFit = benOut, newData = testData)
		blPred  <- postPredict(mibrrFit = blOut, newData = testData)
		
- Posterior predictions can also be generated from `miben` and `mibl` models:

		## Load some data:
		data(predictData)

		missData <- predictData$incomplete
		testData <- predictData$test
		
		## Estimate the imputation models:
		mibenOut <- miben(data = missData)
		miblOut  <- mibl(data = missData)
		
		## Generate out-of-sample predictions:
	    mibenPred <- postPredict(mibrrFit = mibenOut, newData = testData)
		miblPred  <- postPredict(mibrrFit = miblOut, newData = testData)
		
		
[builds]:  https://github.com/kylelang/MIBRR/tree/master/builds/
[docs]:    https://github.com/kylelang/MIBRR/tree/master/documentation/
[src]:     https://github.com/kylelang/MIBRR/tree/master/source/MIBRR
[LICENSE]: https://github.com/kylelang/MIBRR/blob/master/LICENSE


CHANGELOG.md
# Change Log
All notable changes to the **MIBRR** project (and some not so notable ones) will 
be documented in this file.

The format is based on [Keep a Changelog][kacl], and this project (attempts to) 
adhere to [Semantic Versioning][sv].

NOTE: 

- As of `mibrr` version 0.0.0.9000, I am returning to this project after
  approximately 2.5 years away because I was unable to devote any attention to
  this project during my postdoc. All initial development, therefore, is largely
  undocumented.
- On 2017-11-06 the package name was changed from `mibrr` to `MIBRR`, so the 
  version number was reset to 0.0.0.9000, as well

## 0.4.1-9000 - 2020-02-07 - ACTIVE

### Added
- An option to estimate the model without an intercept term

### Fixed 
- A bug with defining the plot boundaries in the `plotLambda` function.

## 0.4.0-9000 - 2019-12-18

### Added
- The ability to run multiple MCEM/Markov chains.
  - The number of chains is controled by the `nChains` argument.
  - A `getLamba` function to extract samples/estimates of the penalty parameters 
	and associated metadata
  - A `plotLambda` function to plot the MCEM trace of the penalty parameters or 
	their loglikelihood
- A so-called "Median Closure Test" as a possible way to assess convergence of 
  MCEM chains.
  - Strong: Checks if the median of each (split) chain is contained within the 
	range of every other (split) chain.
  - Weak: Checks if the overall median of all chains is contained within the 
	range of every (split) chain.
	
### Changed
- The MCEM algorithm is now divided into four phases (as opposed to the previous 
  three phases)
  - This four-phase scheme follows the logic of the MCEM alrogithm described by 
	[Veda, Meng, and Xu (2004)][vedaEtAl:2004].
  - The new phases are:
	1. Approximation: A burn-in phase where the MCEM chain should "forget" its 
	   starting values (equivalent to the Veda et al, 2004, "burn-in" phase)
	1. Tuning: A phase wherein the MC sample size is linearly increased to 
	   narrow in on a precise neighborhood of the MLE (equivalent to the Veda et 
	   al, 2004, "transition" phase)
	1. Stationary: A phase wherein large MC samples are taken to get a reliable 
	   pool of estimates from the MLE's neighborhood (equivalent to the Veda et 
	   al, 2004, "plateau" phase)
	1. Terminal Gibbs Sampling: The average of the Lambda estimates from the 
	   'stationary' phase is used to parameterize the final run of the Gibbs 
	   sampler
- To accomodate the above change, the `iterations` and `sampleSizes` arguments 
  must now be formatted differently.

## 0.3.3.9000 - 2019-12-09

### Fixed
- A bug whereby standardization of Beta was reverted before computing the
  conditional loglikelihood of Lambda (this was causing optimization issues).

### Added
- A control list option, `dumpParamHistory`, to save an RDS file containing the 
  last several iterations of parameter samples. Defaults to `FALSE`.
- A control list option, `phHistoryLength`, that defines how many samples are 
  stored when `dumpParamHistory = TRUE`. Defaults to `10L`.
- A control list option, `centerType`, that defines the central tendency used to
  restart the Gibbs sampler after each MCEM iteration. Possible values are 
  *mode* (the default), *mean*, and *median*.
  
## 0.3.2.9000 - 2019-11-12

### Fixed
- Disabled multithreading in Eigen. We don't want Eigen to autonomously spawn
  its own threads. Doing so leads to unexpected overcommitting of CPU resources
  (especially when running in a cluster environment).
  
### Changed
- Improved the start-up message.

### Added
- A new testing script *visPriors.R* to generate plots visualizing the penalty 
  parameters' prior distributions. This functionality will be incorporated into 
  an exported function in the future.
  
- The option to restart failed optimizations of the penalty parameters in 
  `miben` with randomly perturbed starting values.
	  
	  - The number of restarts is set by the *optMaxRestarts* in the control 
	    list.

- A new control list option, *optStrict*, that dictates whether *optMaxRestarts* 
  failed optimizations of the penalty parameters results in a fatal error or in 
  a warning.
    
## 0.3.1.9000 - 2019-02-26

### Fixed
- A bug with the implementation of the `vanilla` imputation method.

### Changed
- Cleaned up the C++ source code a little bit.

### Added
- A control list option *useBetaMeans* that will tell the Gibbs sampler to use 
  the OLS estimates of *Beta* when updating *sigma* (as opposed to the default 
  option of using *Beta*s most recent sampled values).
- A function, `ppCheck`, to generate overlaid density plots of the observed and 
  posterior predicted samples in a *MibrrFit* object.
  
## 0.3.0.9000 - 2019-01-25

### Changed
- Removed all dependencies except for **mice**; the necessary functions are now
  imported from the previous dependencies.
- Updated the testing script: */testing/testPackage.R* to systematize the 
  testing implemented therein.

### Added
- A new R syntax file: *unitTests.R*.
- Unit tests for the random number samplers (`testMvn`, `testInvGamma`, 
  `testInvGauss`, `testGig`, `testInvChiSq`, `testIncGamma`, and `testSamplers`; 
  not currently exported).
- Unit test for the missing data indexing (`testMissIndex`; not currently 
  exported).
- Unit test for data manipulation/subsetting (`testDataProcessing`; not 
  currently exported).
- Unit test for data scaling (`testDataScaling`; not currently exported).
- Unit test for missing data filling (`testMissFill`; not currently exported).

## 0.2.0.9000 - 2019-01-21

### Fixed
- The scaling issues addressed in the changes listed below seem to have been a 
  large part of the poor performance exhibited in recent simulations. The full 
  extent of the improvement still needs to be evaluated via new simulations.
  
### Changed
- Beta samples are now returned in their raw metric
- Predictors are standardized on-the-fly with respect to the "training set" 

buildPackages.R
### Title:    Build/Install the MIBRR R Package
### Author:   Kyle M. Lang
### Created:  2014-12-07
### Modified: 2019-01-24

rm(list = ls(all = TRUE))

library(RcppEigen)

ver  <- "0.4.1-9000"
prod <- FALSE

## Clean up:
system("rm source/MIBRR/src/RcppExports.cpp \
        rm source/MIBRR/R/RcppExports.R")

## Generate the RccpExports.cpp file:
Rcpp::compileAttributes("source/MIBRR")

if(prod) {
    ## Build the MIBRR package:
    system("R CMD build source/MIBRR")
    
    ## Run CRAN checks:
    system(paste0("R CMD check MIBRR_", ver, ".tar.gz"))
    
    ## Install the MIBRR package:
    install.packages(paste0("MIBRR_", ver, ".tar.gz"),
                     repos = NULL,
                     type  = "source")
    
    ## Move the tar-ball to the builds directory:
    system(paste0("mv MIBRR_", ver, ".tar.gz builds/"))
    
    ## Generate a PDF version of the manual:
    system(paste0("R CMD Rd2pdf --force --output='./documentation/MIBRR.pdf' ",
                  .libPaths()[1],
                  "/MIBRR")
           )
} else {
    system("R CMD INSTALL source/MIBRR")
}
